### PR TITLE
Add --with-launcher-name and --with-product-name configure options

### PR DESCRIPTION
**Pull request by pshipton**
I'll create similar changes for other versions after a review.

### CODE REVIEW

File: make/autoconf/jdk-version.m4 [hunk 1]
```diff
@@ -75,8 +75,6 @@ AC_DEFUN_ONCE([JDKVER_SETUP_JDK_VERSION_NUMBERS],
   . $TOPDIR/closed/autoconf/version-numbers
 
   # Some non-version number information is set in that file
-  AC_SUBST(LAUNCHER_NAME)
-  AC_SUBST(PRODUCT_NAME)
```
Comment (keithc-ca) [inline make/autoconf/jdk-version.m4:79]: This file doesn't need to change to support custom launcher and product names; the customizations belong in `custom-hook.m4`. Setting those variables even as late as in `CUSTOM_LATE_HOOK` is effective.

Comment (pshipton) [reply make/autoconf/jdk-version.m4:79]: I guess that can work for LAUNCHER_NAME, but $PRODUCT_NAMEis used in this file to set the JDK_RC_NAME and needs to have the updated value.

Comment (keithc-ca) [reply make/autoconf/jdk-version.m4:79]: As discussed, moving the customizations is fragile (it doesn't deal well with other references to `LAUNCHER_NAME` or `PRODUCT_NAME`).

[the remark this answers is not in the document]
Comment (pshipton) [reply make/autoconf/jdk-version.m4]: Updated

File: make/autoconf/jdk-version.m4 [hunk 2]
```diff
@@ -109,6 +109,33 @@ AC_DEFUN_ONCE([JDKVER_SETUP_JDK_VERSION_NUMBERS],
   fi
   AC_SUBST(MACOSX_BUNDLE_ID_BASE)
 
+  # The launcher name, if any
+  AC_ARG_WITH(launcher-name, [AS_HELP_STRING([--with-launcher-name],
+      [Set launcher name. Part of the -version and -fullversion output. @<:@not specified@:>@])])
```
Comment (keithc-ca) [inline make/autoconf/jdk-version.m4:114]: Suggested help text:
```
      [Set launcher name (included in -version and -fullversion output). @<:@not specified@:>@])])
```

File: make/autoconf/jdk-version.m4 [hunk 3]
```diff
@@ -109,6 +109,33 @@ AC_DEFUN_ONCE([JDKVER_SETUP_JDK_VERSION_NUMBERS],
...
+  elif test "x$with_launcher_name" != x; then
+    # Only set LAUNCHER_NAME if '--with-launcher-name' was used and is not empty.
+    # Otherwise we will use the value from "version-numbers" included above.
+    LAUNCHER_NAME="$with_launcher_name"
+  fi
+
+  # The product name, if any
+  AC_ARG_WITH(product-name, [AS_HELP_STRING([--with-product-name],
+      [Set product name. Among others, used to set the 'java.runtime.name'
+       system property. @<:@not specified@:>@])])
```
Comment (keithc-ca) [inline make/autoconf/jdk-version.m4:128]: The help test should not have internal line breaks; suggest:
```
      [Set product name. Among other uses, defines the 'java.runtime.name' system property. @<:@not specified@:>@])])
```

File: make/autoconf/jdk-version.m4 [hunk 4]
```diff
@@ -109,6 +109,33 @@ AC_DEFUN_ONCE([JDKVER_SETUP_JDK_VERSION_NUMBERS],
...
   AC_SUBST(MACOSX_BUNDLE_ID_BASE)
 
+  # The launcher name, if any
+  AC_ARG_WITH(launcher-name, [AS_HELP_STRING([--with-launcher-name],
+      [Set launcher name. Part of the -version and -fullversion output. @<:@not specified@:>@])])
+  if test "x$with_launcher_name" = xyes; then
+    AC_MSG_ERROR([--with-launcher-name must have a value])
+  elif [ ! [[ $with_launcher_name =~ ^[[:print:]]*$ ]] ]; then
+    AC_MSG_ERROR([--with-launcher-name contains non-printing characters: $with_launcher_name])
+  elif test "x$with_launcher_name" != x; then
```
Comment (keithc-ca) [inline make/autoconf/jdk-version.m4:119]: This should be
```
  elif test "x$with_launcher_name" != x -a "x$with_launcher_name" != xno; then
```
to properly handle `--without-launcher-name`. Same goes below to handle `--without-product-name`.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: make/autoconf/jdk-version.m4 [hunk 5]
```diff
@@ -109,6 +109,33 @@ AC_DEFUN_ONCE([JDKVER_SETUP_JDK_VERSION_NUMBERS],
   fi
   AC_SUBST(MACOSX_BUNDLE_ID_BASE)
 
+  # The launcher name, if any
+  AC_ARG_WITH(launcher-name, [AS_HELP_STRING([--with-launcher-name],
+      [Set launcher name. Part of the -version and -fullversion output. @<:@not specified@:>@])])
+  if test "x$with_launcher_name" = xyes; then
+    AC_MSG_ERROR([--with-launcher-name must have a value])
+  elif [ ! [[ $with_launcher_name =~ ^[[:print:]]*$ ]] ]; then
```
Comment (keithc-ca) [inline make/autoconf/jdk-version.m4:117]: It would be helpful to have a comment about the need for the outermost brackets (for m4).

Comment (keithc-ca) [reply make/autoconf/jdk-version.m4:117]: Or not, I see the same pattern in the handler for `--with-macosx-bundle-id-base`.

Comment (pshipton) [reply make/autoconf/jdk-version.m4:117]: I just copied this stuff from vendor-name. What would you like the comment to be?

Comment (pshipton) [reply make/autoconf/jdk-version.m4:117]: Ah, didn't see your update until now.

Comment (pshipton) [reply make/autoconf/jdk-version.m4:117]: All the review comments should be addressed then.

File: make/autoconf/jdk-version.m4 (hunk 3 above)
Comment (keithc-ca) [reply make/autoconf/jdk-version.m4]: The line break was removed, but the wording was not updated.

Comment (pshipton) [reply make/autoconf/jdk-version.m4]: Updated now.

**Comment by keithc-ca** [comment]
Jenkins compile osx jdk11

**Comment by keithc-ca** [comment]
Like the jdk8 build, it succeeded, but the upload to artifactory failed.

**Review by keithc-ca — APPROVED** [review]

**Comment by pshipton** [comment]
Building with `--with-launcher-name=peter --with-product-name=Peter`
```
14:49:47  peter version "11.0.12-internal" 2021-07-20
14:49:47  Peter Runtime Environment (build 11.0.12-internal+0-adhoc.jenkins.BuildJDK11s390xlinuxPersonal)
14:49:47  Eclipse OpenJ9 VM (build custom-5a258bf3b4e, JRE 11 Linux s390x-64-Bit Compressed References 20210707_116 (JIT enabled, AOT enabled)
14:49:47  OpenJ9   - 5a258bf3b4e
14:49:47  OMR      - 42d7f624612
14:49:47  JCL      - d803c146923 based on jdk-11.0.12+6)
```

**Pull request merged**